File: libs/sdk/src/lib/resources/agents.ts
import type { HttpClient } from '../http-client';
import type { ApiResponse, RequestOptions } from '../types';
import type {
  AgentRole,
  AgentStatus,
  Proficiency,
} from '../shared-types';

/**
 * Agent-related types
 */
export interface Agent {
  id: string;
  agentId: string;
  name: string;
  role: AgentRole;
  level: number;
  model?: string;
  status: AgentStatus;
  currentBalance: number;
  managementFeePct?: number;
  budgetPeriodLimit?: number;
  budgetPeriodSpent?: number;
  capabilities?: AgentCapability[];
  metadata?: Record<string, unknown>;
  parentId?: string;
  createdAt: string;
  updatedAt?: string;
}

export interface AgentCapability {
  id: string;
  capability: string;
  proficiency: Proficiency;
}

export interface CreateAgentDto {
  name: string;
  role: AgentRole;
  level: number;
  model?: string;
}

export interface UpdateAgentDto {
  name?: string;
  status?: AgentStatus;
  metadata?: Record<string, unknown>;
}

export interface SpawnAgentDto {
  name: string;
  level: number;
  capabilities?: Array<{ capability: string; proficiency: Proficiency }>;
}

export interface ReputationSummary {
  agentId: string;
  reputationScore: number;
  level: string;
  tasksCompleted: number;
  successRate: number;
}

export interface BudgetStatus {
  agentId: string;
  currentBalance: number;
  budgetPeriodLimit?: number;
  budgetPeriodSpent: number;
  budgetRemaining?: number;
}

/**
 * Agents Resource
 */
export class AgentsResource {
  constructor(private http: HttpClient) {}

  /**
   * List all agents in the organization
   */
  async list(options?: RequestOptions): Promise<Agent[]> {
    const response = await this.http.get<ApiResponse<Agent[]>>(
      '/agents',
      options
    );
    return response.data;
  }

  /**
   * Get a specific agent by ID
   */
  async get(id: string, options?: RequestOptions): Promise<Agent> {
    const response = await this.http.get<ApiResponse<Agent>>(
      `/agents/${id}`,
      options
    );
    return response.data;
  }

  /**
   * Create a new agent (requires HR role)
   */
  async create(
    data: CreateAgentDto,
    options?: RequestOptions
  ): Promise<{ agent: Agent; secret: string }> {
    const response = await this.http.post<
      ApiResponse<Agent> & { secret: string }
    >('/agents/register', data, options);
    return {
      agent: response.data,
      secret: response.secret,
    };
  }

  /**
   * Update an agent (requires HR role)
   */
  async update(
    id: string,
    data: UpdateAgentDto,
    options?: RequestOptions
  ): Promise<Agent> {
    const response = await this.http.patch<ApiResponse<Agent>>(
      `/agents/${id}`,
      data,
      options
    );
    return response.data;
  }

  /**
   * Revoke an agent (requires HR role)
   */
  async revoke(
    id: string,
    options?: RequestOptions
  ): Promise<{ id: string; status: AgentStatus }> {
    const response = await this.http.post<
      ApiResponse<{ id: string; status: AgentStatus }>
    >(`/agents/${id}/revoke`, undefined, options);
    return response.data;
  }

  /**
   * Spawn a new child agent
   */
  async spawn(
    data: SpawnAgentDto,
    options?: RequestOptions
  ): Promise<{ agent: Agent; secret: string }> {
    const response = await this.http.post<
      ApiResponse<Agent> & { secret: string }
    >('/agents/spawn', data, options);
    return {
      agent: response.data,
      secret: response.secret,
    };
  }

  /**
   * Activate a pending agent
   */
  async activate(
    id: string,
    options?: RequestOptions
  ): Promise<{ id: string; status: AgentStatus }> {
    const response = await this.http.post<
      ApiResponse<{ id: string; status: AgentStatus }>
    >(`/agents/${id}/activate`, undefined, options);
    return response.data;
  }

  /**
   * Reject a pending agent
   */
  async reject(
    id: string,
    reason?: string,
    options?: RequestOptions
  ): Promise<void> {
    await this.http.delete(`/agents/${id}/reject`, options);
  }

  /**
   * Get pending agents awaiting activation
   */
  async getPending(options?: RequestOptions): Promise<Agent[]> {
    const response = await this.http.get<ApiResponse<Agent[]>>(
      '/agents/pending',
      options
    );
    return response.data;
  }

  /**
   * Get agent hierarchy
   */
  async getHierarchy(
    id: string,
    depth = 3,
    options?: RequestOptions
  ): Promise<unknown> {
    const response = await this.http.get<ApiResponse<unknown>>(
      `/agents/${id}/hierarchy?depth=${depth}`,
      options
    );
    return response.data;
  }

  /**
   * Get credit balance for an agent
   */
  async getBalance(id: string, options?: RequestOptions): Promise<number> {
    const response = await this.http.get<ApiResponse<{ balance: number }>>(
      `/agents/${id}/credits/balance`,
      options
    );
    return response.data.balance;
  }

  /**
   * Get budget status
   */
  async getBudgetStatus(
    id: string,
    options?: RequestOptions
  ): Promise<BudgetStatus> {
    const response = await this.http.get<ApiResponse<BudgetStatus>>(
      `/agents/${id}/budget`,
      options
    );
    return response.data;
  }

  /**
   * Get reputation summary
   */
  async getReputation(
    id: string,
    options?: RequestOptions
  ): Promise<ReputationSummary> {
    const response = await this.http.get<ApiResponse<ReputationSummary>>(
      `/agents/${id}/reputation`,
      options
    );
    return response.data;
  }

  /**
   * Award quality bonus
   */
  async awardBonus(
    id: string,
    reason: string,
    amount?: number,
    options?: RequestOptions
  ): Promise<unknown> {
    const response = await this.http.post<ApiResponse<unknown>>(
      `/agents/${id}/reputation/bonus`,
      { reason, amount },
      options
    );
    return response.data;
  }

  /**
   * Apply quality penalty
   */
  async applyPenalty(
    id: string,
    reason: string,
    amount?: number,
    options?: RequestOptions
  ): Promise<unknown> {
    const response = await this.http.post<ApiResponse<unknown>>(
      `/agents/${id}/reputation/penalty`,
      { reason, amount },
      options
    );
    return response.data;
  }

  /**
   * Add a capability to an agent
   */
  async addCapability(
    id: string,
    capability: string,
    proficiency: Proficiency,
    options?: RequestOptions
  ): Promise<AgentCapability> {
    const response = await this.http.post<ApiResponse<AgentCapability>>(
      `/agents/${id}/capabilities`,
      { capability, proficiency },
      options
    );
    return response.data;
  }

  /**
   * Find agents with specific capabilities
   */
  async findByCapabilities(
    capabilities: string[],
    minProficiency?: Proficiency,
    options?: RequestOptions
  ): Promise<Agent[]> {
    const params = new URLSearchParams({
      capabilities: capabilities.join(','),
    });
    if (minProficiency) {
      params.set('minProficiency', minProficiency);
    }
    const response = await this.http.get<ApiResponse<Agent[]>>(
      `/agents/capabilities/match?${params.toString()}`,
      options
    );
    return response.data;
  }
}

File: sdks/python/src/openspawn/resources/agents.py
"""Agents resource module."""

from typing import Any, Optional

from openspawn.http_client import AsyncHTTPClient, SyncHTTPClient
from openspawn.models import Agent, CreateAgentRequest, UpdateAgentRequest


class AgentsResource:
    """Synchronous agents resource."""

    def __init__(self, client: SyncHTTPClient) -> None:
        self._client = client

    def list(self) -> list[Agent]:
        """
        List all agents in the organization.

        Returns:
            List of agents
        """
        response = self._client.request("GET", "/agents")
        return [Agent(**agent) for agent in response["data"]]

    def get(self, agent_id: str) -> Agent:
        """
        Get a specific agent by ID.

        Args:
            agent_id: Agent ID

        Returns:
            Agent instance
        """
        response = self._client.request("GET", f"/agents/{agent_id}")
        return Agent(**response["data"])

    def create(self, request: CreateAgentRequest) -> dict[str, Any]:
        """
        Register a new agent (HR role required).

        Args:
            request: Agent creation request

        Returns:
            Response containing agent data and secret (SAVE THE SECRET!)
        """
        response = self._client.request(
            "POST",
            "/agents/register",
            json_data=request.model_dump(exclude_none=True),
            idempotent=True,
        )
        return response

    def update(self, agent_id: str, request: UpdateAgentRequest) -> Agent:
        """
        Update an agent (HR role required).

        Args:
            agent_id: Agent ID
            request: Agent update request

        Returns:
            Updated agent
        """
        response = self._client.request(
            "PATCH",
            f"/agents/{agent_id}",
            json_data=request.model_dump(exclude_none=True),
        )
        return Agent(**response["data"])

    def revoke(self, agent_id: str) -> dict[str, Any]:
        """
        Revoke an agent (HR role required).

        Args:
            agent_id: Agent ID

        Returns:
            Response containing revoked agent status
        """
        response = self._client.request("POST", f"/agents/{agent_id}/revoke")
        return response

    def get_balance(self, agent_id: str) -> int:
        """
        Get credit balance for an agent.

        Args:
            agent_id: Agent ID

        Returns:
            Current balance
        """
        response = self._client.request("GET", f"/agents/{agent_id}/credits/balance")
        return response["data"]

    def spawn(
        self,
        name: str,
        level: int,
        *,
        model: Optional[str] = None,
        metadata: Optional[dict[str, Any]] = None,
    ) -> dict[str, Any]:
        """
        Spawn a new child agent.

        Args:
            name: Agent name
            level: Agent level
            model: AI model name
            metadata: Additional metadata

        Returns:
            Response containing spawned agent and secret
        """
        response = self._client.request(
            "POST",
            "/agents/spawn",
            json_data={
                "name": name,
                "level": level,
                "model": model,
                "metadata": metadata,
            },
            idempotent=True,
        )
        return response

    def activate(self, agent_id: str) -> dict[str, Any]:
        """
        Activate a pending agent.

        Args:
            agent_id: Agent ID

        Returns:
            Response with activation status
        """
        response = self._client.request("POST", f"/agents/{agent_id}/activate")
        return response

    def reject(self, agent_id: str, reason: Optional[str] = None) -> dict[str, Any]:
        """
        Reject a pending agent.

        Args:
            agent_id: Agent ID
            reason: Rejection reason

        Returns:
            Response message
        """
        response = self._client.request(
            "DELETE",
            f"/agents/{agent_id}/reject",
            json_data={"reason": reason} if reason else {},
        )
        return response

    def get_hierarchy(self, agent_id: str, depth: int = 3) -> dict[str, Any]:
        """
        Get agent hierarchy.

        Args:
            agent_id: Agent ID
            depth: Hierarchy depth

        Returns:
            Hierarchy data
        """
        response = self._client.request(
            "GET",
            f"/agents/{agent_id}/hierarchy",
            params={"depth": str(depth)},
        )
        return response["data"]


class AsyncAgentsResource:
    """Asynchronous agents resource."""

    def __init__(self, client: AsyncHTTPClient) -> None:
        self._client = client

    async def list(self) -> list[Agent]:
        """
        List all agents in the organization.

        Returns:
            List of agents
        """
        response = await self._client.request("GET", "/agents")
        return [Agent(**agent) for agent in response["data"]]

    async def get(self, agent_id: str) -> Agent:
        """
        Get a specific agent by ID.

        Args:
            agent_id: Agent ID

        Returns:
            Agent instance
        """
        response = await self._client.request("GET", f"/agents/{agent_id}")
        return Agent(**response["data"])

    async def create(self, request: CreateAgentRequest) -> dict[str, Any]:
        """
        Register a new agent (HR role required).

        Args:
            request: Agent creation request

        Returns:
            Response containing agent data and secret (SAVE THE SECRET!)
        """
        response = await self._client.request(
            "POST",
            "/agents/register",
            json_data=request.model_dump(exclude_none=True),
            idempotent=True,
        )
        return response

    async def update(self, agent_id: str, request: UpdateAgentRequest) -> Agent:
        """
        Update an agent (HR role required).

        Args:
            agent_id: Agent ID
            request: Agent update request

        Returns:
            Updated agent
        """
        response = await self._client.request(
            "PATCH",
            f"/agents/{agent_id}",
            json_data=request.model_dump(exclude_none=True),
        )
        return Agent(**response["data"])

    async def revoke(self, agent_id: str) -> dict[str, Any]:
        """
        Revoke an agent (HR role required).

        Args:
            agent_id: Agent ID

        Returns:
            Response containing revoked agent status
        """
        response = await self._client.request("POST", f"/agents/{agent_id}/revoke")
        return response

    async def get_balance(self, agent_id: str) -> int:
        """
        Get credit balance for an agent.

        Args:
            agent_id: Agent ID

        Returns:
            Current balance
        """
        response = await self._client.request("GET", f"/agents/{agent_id}/credits/balance")
        return response["data"]

    async def spawn(
        self,
        name: str,
        level: int,
        *,
        model: Optional[str] = None,
        metadata: Optional[dict[str, Any]] = None,
    ) -> dict[str, Any]:
        """
        Spawn a new child agent.

        Args:
            name: Agent name
            level: Agent level
            model: AI model name
            metadata: Additional metadata

        Returns:
            Response containing spawned agent and secret
        """
        response = await self._client.request(
            "POST",
            "/agents/spawn",
            json_data={
                "name": name,
                "level": level,
                "model": model,
                "metadata": metadata,
            },
            idempotent=True,
        )
        return response

    async def activate(self, agent_id: str) -> dict[str, Any]:
        """
        Activate a pending agent.

        Args:
            agent_id: Agent ID

        Returns:
            Response with activation status
        """
        response = await self._client.request("POST", f"/agents/{agent_id}/activate")
        return response

    async def reject(self, agent_id: str, reason: Optional[str] = None) -> dict[str, Any]:
        """
        Reject a pending agent.

        Args:
            agent_id: Agent ID
            reason: Rejection reason

        Returns:
            Response message
        """
        response = await self._client.request(
            "DELETE",
            f"/agents/{agent_id}/reject",
            json_data={"reason": reason} if reason else {},
        )
        return response

    async def get_hierarchy(self, agent_id: str, depth: int = 3) -> dict[str, Any]:
        """
        Get agent hierarchy.

        Args:
            agent_id: Agent ID
            depth: Hierarchy depth

        Returns:
            Hierarchy data
        """
        response = await self._client.request(
            "GET",
            f"/agents/{agent_id}/hierarchy",
            params={"depth": str(depth)},
        )
        return response["data"]
